tensorflow_federated/python/core/impl/executors/federating_executor.py

```diff
@@ -453,6 +451,221 @@ def validate_executor_placements(cls, executor_placements):
+    return await self._zip(arg, placement_literals.CLIENTS, all_equal=False)
+
+  async def federated_reduce(self, arg):
+    self._check_arg_is_anonymous_tuple(arg)
+    if len(arg.internal_representation) != 3:
+      raise ValueError(
+          'Expected 3 elements in the `federated_reduce()` argument tuple, '
+          'found {}.'.format(len(arg.internal_representation)))
+
+    val_type = arg.type_signature[0]
+    py_typecheck.check_type(val_type, computation_types.FederatedType)
+    item_type = val_type.member
+    zero_type = arg.type_signature[1]
+    op_type = arg.type_signature[2]
+    type_utils.check_equivalent_types(
+        op_type, type_factory.reduction_op(zero_type, item_type))
+
+    val = arg.internal_representation[0]
+    py_typecheck.check_type(val, list)
+    aggr = self._get_child_executors(placement_literals.AGGREGATOR, index=0)
+    server = self._get_child_executors(placement_literals.SERVER, index=0)
+
+    async def _move(v, target_executor):
+      return await target_executor.create_value(await v.compute(), item_type)
+
+    aggregands = await asyncio.gather(*[_move(v, aggr) for v in val])
+
+    zero = await aggr.create_value(
+        await (await self.federating_executor.create_selection(
+            arg, index=1)).compute(),
+        zero_type)
+    op = await aggr.create_value(arg.internal_representation[2], op_type)
+
+    result = zero
+    for item in aggregands:
+      result = await aggr.create_call(
+          op, await aggr.create_tuple(
+              anonymous_tuple.AnonymousTuple([(None, result), (None, item)])))
+    return FederatingExecutorValue([result],
+                                   computation_types.FederatedType(
+                                       result.type_signature,
+                                       placement_literals.SERVER,
+                                       all_equal=True))
+
+  async def federated_aggregate(self, arg):
+    val_type, zero_type, accumulate_type, _, report_type = (
+        executor_utils.parse_federated_aggregate_argument_types(
+            arg.type_signature))
+    py_typecheck.check_type(arg.internal_representation,
+                            anonymous_tuple.AnonymousTuple)
+    py_typecheck.check_len(arg.internal_representation, 5)
+
+    # Note: This is a simple initial implementation that simply forwards this
+    # to `federated_reduce()`. The more complete implementation would be able
+    # to take advantage of the parallelism afforded by `merge` to reduce the
+    # cost from liner (with respect to the number of clients) to sub-linear.
+
+    # TODO(b/134543154): Expand this implementation to take advantage of the
+    # parallelism afforded by `merge`.
+    fed_ex = self.federating_executor
+
+    val = arg.internal_representation[0]
+    zero = arg.internal_representation[1]
+    accumulate = arg.internal_representation[2]
+    pre_report = await fed_ex._compute_intrinsic_federated_reduce(
```

Comment:
In the `TrustedAggregatorIntrinsicStrategy` in several places, why are we using `._compute_intrinsic_federated_reduce` from the federating executor instead of `self.federated_reduce`?

Reply:
Because it's standard to trace _compute_intrinsic, and so calling self.federated_x from inside self.federated_y would mean losing the trace for x

Reply:
ah my bad, forgot this for a sec:
```
async def _compute_intrinsic_federated_reduce(self, arg):
    return await self.intrinsic_strategy.federated_reduce(arg)
```